modules/src/applications/transfer/denom.rs

```diff
@@ -0,0 +1,628 @@
+    type Error = Error;
+
+    fn try_from(v: Vec<&'a str>) -> Result<Self, Self::Error> {
+        if v.len() % 2 != 0 {
+            return Err(Error::invalid_trace_length(v.len()));
+        }
+
+        let mut trace = vec![];
+        let id_pairs = v.windows(2).map(|paths| (paths[0], paths[1]));
+        for (pos, (port_id, channel_id)) in id_pairs.step_by(2).enumerate() {
+            let port_id =
+                PortId::from_str(port_id).map_err(|e| Error::invalid_trace_port_id(pos, e))?;
+            let channel_id = ChannelId::from_str(channel_id)
+                .map_err(|e| Error::invalid_trace_channel_id(pos, e))?;
+            trace.push(TracePrefix {
+                port_id,
+                channel_id,
+            });
+        }
+
+        Ok(trace.into())
+    }
+}
+
+impl FromStr for TracePath {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts = {
+            let parts: Vec<&str> = s.split('/').collect();
+            if parts.len() == 1 && parts[0].trim().is_empty() {
+                vec![]
+            } else {
+                parts
+            }
+        };
+        parts.try_into()
+    }
+}
+
+impl fmt::Display for TracePath {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let path = self
+            .0
+            .iter()
+            .map(|prefix| prefix.to_string())
+            .collect::<Vec<String>>()
+            .join("/");
+        write!(f, "{}", path)
+    }
+}
+
+/// Indicates whether the sender chain is acting as a source or sink. Each send to any chain other
+/// than the one it was previously received from is a movement forwards in the token's timeline. In
+/// these instances the sender chain is acting as the source zone.
+pub enum Source {
+    Sender,
+    Receiver,
+}
+
+/// A type that contains the base denomination for ICS20 and the source tracing information path.
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct DenomTrace {
+    /// A series of `{port-id}/{channel-id}`s for tracing the source of the token.
+    #[serde(with = "serde_string")]
+    trace_path: TracePath,
+    /// Base denomination of the relayed fungible token.
+    base_denom: Denom,
+}
+
+impl DenomTrace {
+    /// Returns a coin denomination for an ICS20 fungible token in the format
+    /// 'ibc/{Hash(trace_path/base_denom)}'.
+    pub fn hashed(&self) -> HashedDenom {
+        HashedDenom::from(self)
+    }
+
+    /// Returns true iff this path has the specified prefix
+    pub fn has_prefix(&self, prefix: &TracePrefix) -> bool {
+        self.trace_path
+            .0
+            .first()
+            .map(|p| p == prefix)
+            .unwrap_or(false)
+    }
+
+    /// Removes the specified prefix from the `trace_path` if there is a match.
+    pub fn remove_prefix(&mut self, prefix: &TracePrefix) {
+        if self.has_prefix(prefix) {
+            self.trace_path.0.drain(..0);
```

Comment:
Can we avoid using `drain()` since the interface is confusing, as previously discussed? Maybe we can change the order of the prefixes in `TracePath` such that removing a prefix can be achieved using `pop()`?